File: lib/ticket_dispenser.ex
defmodule TicketDispenser do
	use GenServer

	def start() do
		GenServer.start(__MODULE__, 1, [])
	end

	def stop(server) do
		GenServer.call(server, :stop)
	end

	def take(server) do
		n = GenServer.call(server, :read)
		GenServer.call(server, {:write, n+1})
	end
	
	def reset(server) do
		GenServer.call(server, {:write, 1})
	end

	
	## server callbacks
	
	def init(n) do
    {:ok, n}
  end

  def handle_call(:read, _from, n) do
    {:reply, n, n}
  end
	def handle_call({:write, m}, _from, n) do
    {:reply, n, m}
  end
	def handle_call(:stop, _from, n) do
    {:stop, :normal, n, n}
	end

end

File: src/ticket_dispenser.erl
-module(ticket_dispenser).

-export([start/0, stop/1, reset/1, take/1]).

start() ->
  {ok, spawn(fun() -> run(0) end)}.

stop(Pid) ->
  Pid ! {stop, self()},
  ok.

reset(Pid) ->
  write(Pid, 0).

take(Pid) ->
  N = read(Pid),
  N2 = write(Pid, N + 1),
  N2.

% -------------------

read(Pid) ->
  Pid ! {read, self()},
  receive N -> N end.

write(Pid, N) ->
  Pid ! {write, self(), N},
  receive N1 -> N1 end.

run(N) ->
  receive 
    {read, From} ->
      From ! N,
      run(N);
    {write, From, NewN} ->
      From ! NewN,
      run(NewN);
    {stop, _From} ->
      ok;
    Input ->
      io:format("Unexpected input ~p~n", [Input]),
      run(N)
  end.


